File: ios/Classes/Performance/PerformanceCollector.swift
import Foundation
import CoreMedia

/// iOS 性能数据收集器
///
/// 收集 ijkplayer 播放过程中的性能指标，包括：
/// - 帧率 (FPS)
/// - 丢帧数
/// - 解码耗时
/// - 缓冲区状态
public class PerformanceCollector {
    
    // MARK: - Properties
    
    private weak var mediaPlayer: IJKFFMoviePlayerController?
    private var timer: Timer?
    private var isCollecting = false
    
    // 帧率计算
    private var frameCount: Int64 = 0
    private var lastFpsCalcTime: CFAbsoluteTime = 0
    private var currentFps: Double = 0
    
    // 丢帧统计
    private var droppedFrames: Int64 = 0
    private var decodedFrames: Int64 = 0
    
    // 解码耗时
    private var videoDecodeTimeMs: Double = 0
    private var audioDecodeTimeMs: Double = 0
    private var renderTimeMs: Double = 0
    
    // 采样配置
    private var sampleIntervalMs: Int = 1000
    private var collectBandwidth = true
    private var collectBitrate = true
    
    // 回调
    public var callback: PerformanceCallback?
    
    // MARK: - Initialization
    
    public init() {}
    
    // MARK: - Public Methods
    
    /// 绑定播放器
    public func attach(player: IJKFFMoviePlayerController) {
        mediaPlayer = player
    }
    
    /// 解绑播放器
    public func detach() {
        stop()
        mediaPlayer = nil
    }
    
    /// 配置采样参数
    public func configure(
        sampleIntervalMs: Int = 1000,
        collectBandwidth: Bool = true,
        collectBitrate: Bool = true
    ) {
        self.sampleIntervalMs = sampleIntervalMs
        self.collectBandwidth = collectBandwidth
        self.collectBitrate = collectBitrate
    }
    
    /// 开始收集
    public func start() {
        guard !isCollecting else { return }
        isCollecting = true
        
        lastFpsCalcTime = CFAbsoluteTimeGetCurrent()
        frameCount = 0
        
        scheduleCollection()
    }
    
    /// 停止收集
    public func stop() {
        isCollecting = false
        timer?.invalidate()
        timer = nil
    }
    
    /// 记录一帧
    public func onFrameRendered() {
        frameCount += 1
    }
    
    /// 记录丢帧
    public func onFrameDropped() {
        droppedFrames += 1
    }
    
    /// 记录解码完成
    public func onFrameDecoded(videoTimeMs: Double, audioTimeMs: Double) {
        decodedFrames += 1
        videoDecodeTimeMs = videoTimeMs
        audioDecodeTimeMs = audioTimeMs
    }
    
    /// 记录渲染耗时
    public func onRenderComplete(timeMs: Double) {
        renderTimeMs = timeMs
    }
    
    /// 获取当前性能指标
    public func getMetrics() -> PerformanceMetrics {
        // 计算帧率
        let now = CFAbsoluteTimeGetCurrent()
        let elapsed = now - lastFpsCalcTime
        if elapsed > 0 {
            currentFps = Double(frameCount) / elapsed
        }
        
        // 从 ijkplayer 获取额外信息
        var bufferLengthMs = 0
        var videoBitrate: Int?
        var audioBitrate: Int?
        var width: Int?
        var height: Int?
        var isHardwareDecoding = false
        var bandwidthBps: Int?
        
        if let player = mediaPlayer {
            // 获取视频信息
            let videoSize = player.naturalSize
            width = Int(videoSize.width)
            height = Int(videoSize.height)
            
            // 获取缓冲信息
            if let monitor = player.monitor {
                let videoCachedDuration = monitor.videoCachedDuration
                let audioCachedDuration = monitor.audioCachedDuration
                bufferLengthMs = Int(max(videoCachedDuration, audioCachedDuration) * 1000)
                
                // 获取解码器信息
                isHardwareDecoding = player.isVideoToolboxOpen
                
                // 获取码率信息
                if collectBitrate {
                    videoBitrate = Int(monitor.tcpSpeed)
                }
                
                // 获取带宽估计
                if collectBandwidth {
                    bandwidthBps = Int(monitor.tcpSpeed)
                }
            }
        }
        
        return PerformanceMetrics(
            frameRate: currentFps,
            droppedFrames: Int(droppedFrames),
            decodedFrames: Int(decodedFrames),
            videoDecodeTimeMs: videoDecodeTimeMs,
            audioDecodeTimeMs: audioDecodeTimeMs,
            renderTimeMs: renderTimeMs,
            bufferLengthMs: bufferLengthMs,
            bandwidthBps: bandwidthBps,
            isHardwareDecoding: isHardwareDecoding,
            videoBitrate: videoBitrate,
            audioBitrate: audioBitrate,
            width: width,
            height: height
        )
    }
    
    /// 转换为 Dictionary（用于 Platform Channel）
    public func getMetricsMap() -> [String: Any?] {
        return getMetrics().toMap()
    }
    
    // MARK: - Private Methods
    
    private func scheduleCollection() {
        guard isCollecting else { return }
        
        let interval = TimeInterval(sampleIntervalMs) / 1000.0
        timer = Timer.scheduledTimer(withTimeInterval: interval, repeats: true) { [weak self] _ in
            guard let self = self, self.isCollecting else { return }
            
            let metrics = self.getMetrics()
            self.callback?.onPerformanceUpdate(metrics: metrics)
            
            // 重置帧计数
            self.frameCount = 0
            self.lastFpsCalcTime = CFAbsoluteTimeGetCurrent()
        }
    }
}

// MARK: - PerformanceCallback Protocol

public protocol PerformanceCallback: AnyObject {
    func onPerformanceUpdate(metrics: PerformanceMetrics)
}

// MARK: - PerformanceMetrics

public struct PerformanceMetrics {
    public let frameRate: Double
    public let droppedFrames: Int
    public let decodedFrames: Int
    public let videoDecodeTimeMs: Double
    public let audioDecodeTimeMs: Double
    public let renderTimeMs: Double
    public let bufferLengthMs: Int
    public let bandwidthBps: Int?
    public let isHardwareDecoding: Bool
    public let videoBitrate: Int?
    public let audioBitrate: Int?
    public let width: Int?
    public let height: Int?
    
    public func toMap() -> [String: Any?] {
        return [
            "frameRate": frameRate,
            "droppedFrames": droppedFrames,
            "decodedFrames": decodedFrames,
            "videoDecodeTimeMs": videoDecodeTimeMs,
            "audioDecodeTimeMs": audioDecodeTimeMs,
            "renderTimeMs": renderTimeMs,
            "bufferLengthMs": bufferLengthMs,
            "bandwidthBps": bandwidthBps,
            "isHardwareDecoding": isHardwareDecoding,
            "videoBitrate": videoBitrate,
            "audioBitrate": audioBitrate,
            "width": width,
            "height": height
        ]
    }
}

// MARK: - IJKFFMoviePlayerController Extensions (Placeholder)

/// 这是一个占位类型声明，实际项目中应使用真实的 IJKFFMoviePlayerController
/// 这里仅作为编译时的类型占位
#if !IJKPLAYER_AVAILABLE
public class IJKFFMoviePlayerController: NSObject {
    public var naturalSize: CGSize { return .zero }
    public var isVideoToolboxOpen: Bool { return false }
    public var monitor: IJKFFMonitor? { return nil }
}

public class IJKFFMonitor: NSObject {
    public var videoCachedDuration: Double { return 0 }
    public var audioCachedDuration: Double { return 0 }
    public var tcpSpeed: Double { return 0 }
}
#endif

File: android/src/main/kotlin/com/example/mango_player/performance/PerformanceCollector.kt
package com.example.mango_player.performance

import android.os.Handler
import android.os.Looper
import android.os.SystemClock
import tv.danmaku.ijk.media.player.IMediaPlayer
import java.util.concurrent.atomic.AtomicBoolean
import java.util.concurrent.atomic.AtomicLong

/**
 * Android 性能数据收集器
 *
 * 收集 ijkplayer 播放过程中的性能指标，包括：
 * - 帧率 (FPS)
 * - 丢帧数
 * - 解码耗时
 * - 缓冲区状态
 */
class PerformanceCollector {
    
    private var mediaPlayer: IMediaPlayer? = null
    private val handler = Handler(Looper.getMainLooper())
    private val isCollecting = AtomicBoolean(false)
    
    // 帧率计算
    private var frameCount = AtomicLong(0)
    private var lastFrameTime = AtomicLong(0)
    private var lastFpsCalcTime = AtomicLong(0)
    private var currentFps = 0.0
    
    // 丢帧统计
    private var droppedFrames = AtomicLong(0)
    private var decodedFrames = AtomicLong(0)
    
    // 解码耗时
    private var videoDecodeTimeMs = 0.0
    private var audioDecodeTimeMs = 0.0
    private var renderTimeMs = 0.0
    
    // 采样配置
    private var sampleIntervalMs = 1000L
    private var collectBandwidth = true
    private var collectBitrate = true
    
    // 回调
    private var callback: PerformanceCallback? = null
    
    /**
     * 绑定播放器
     */
    fun attach(player: IMediaPlayer) {
        mediaPlayer = player
    }
    
    /**
     * 解绑播放器
     */
    fun detach() {
        stop()
        mediaPlayer = null
    }
    
    /**
     * 设置性能回调
     */
    fun setCallback(callback: PerformanceCallback?) {
        this.callback = callback
    }
    
    /**
     * 配置采样参数
     */
    fun configure(
        sampleIntervalMs: Long = 1000,
        collectBandwidth: Boolean = true,
        collectBitrate: Boolean = true
    ) {
        this.sampleIntervalMs = sampleIntervalMs
        this.collectBandwidth = collectBandwidth
        this.collectBitrate = collectBitrate
    }
    
    /**
     * 开始收集
     */
    fun start() {
        if (isCollecting.getAndSet(true)) return
        
        lastFpsCalcTime.set(SystemClock.elapsedRealtime())
        frameCount.set(0)
        
        scheduleCollection()
    }
    
    /**
     * 停止收集
     */
    fun stop() {
        isCollecting.set(false)
        handler.removeCallbacksAndMessages(null)
    }
    
    /**
     * 记录一帧
     */
    fun onFrameRendered() {
        frameCount.incrementAndGet()
        lastFrameTime.set(SystemClock.elapsedRealtime())
    }
    
    /**
     * 记录丢帧
     */
    fun onFrameDropped() {
        droppedFrames.incrementAndGet()
    }
    
    /**
     * 记录解码完成
     */
    fun onFrameDecoded(videoTimeMs: Double, audioTimeMs: Double) {
        decodedFrames.incrementAndGet()
        videoDecodeTimeMs = videoTimeMs
        audioDecodeTimeMs = audioTimeMs
    }
    
    /**
     * 记录渲染耗时
     */
    fun onRenderComplete(timeMs: Double) {
        renderTimeMs = timeMs
    }
    
    /**
     * 获取当前性能指标
     */
    fun getMetrics(): PerformanceMetrics {
        val player = mediaPlayer
        
        // 计算帧率
        val now = SystemClock.elapsedRealtime()
        val elapsed = now - lastFpsCalcTime.get()
        if (elapsed > 0) {
            currentFps = frameCount.get() * 1000.0 / elapsed
        }
        
        // 从 ijkplayer 获取额外信息
        var bufferLengthMs = 0
        var videoBitrate: Int? = null
        var audioBitrate: Int? = null
        var width: Int? = null
        var height: Int? = null
        var isHardwareDecoding = false
        var bandwidthBps: Int? = null
        
        player?.let { p ->
            try {
                // 尝试获取视频信息
                width = p.videoWidth
                height = p.videoHeight
                
                // ijkplayer 特定信息获取
                if (p is tv.danmaku.ijk.media.player.IjkMediaPlayer) {
                    // 获取缓冲区信息
                    val videoCachedDuration = p.videoCachedDuration
                    val audioCachedDuration = p.audioCachedDuration
                    bufferLengthMs = maxOf(videoCachedDuration, audioCachedDuration).toInt()
                    
                    // 获取解码器信息
                    val videoDecoder = p.videoDecoder
                    isHardwareDecoding = videoDecoder?.contains("mediacodec", ignoreCase = true) == true
                    
                    // 获取码率信息
                    if (collectBitrate) {
                        videoBitrate = p.tcpSpeed.toInt()
                    }
                    
                    // 获取带宽估计
                    if (collectBandwidth) {
                        bandwidthBps = p.tcpSpeed.toInt()
                    }
                }
            } catch (e: Exception) {
                // 忽略获取信息时的错误
            }
        }
        
        return PerformanceMetrics(
            frameRate = currentFps,
            droppedFrames = droppedFrames.get().toInt(),
            decodedFrames = decodedFrames.get().toInt(),
            videoDecodeTimeMs = videoDecodeTimeMs,
            audioDecodeTimeMs = audioDecodeTimeMs,
            renderTimeMs = renderTimeMs,
            bufferLengthMs = bufferLengthMs,
            bandwidthBps = bandwidthBps,
            isHardwareDecoding = isHardwareDecoding,
            videoBitrate = videoBitrate,
            audioBitrate = audioBitrate,
            width = width,
            height = height
        )
    }
    
    /**
     * 转换为 Map（用于 Platform Channel）
     */
    fun getMetricsMap(): Map<String, Any?> {
        return getMetrics().toMap()
    }
    
    private fun scheduleCollection() {
        if (!isCollecting.get()) return
        
        handler.postDelayed({
            if (isCollecting.get()) {
                val metrics = getMetrics()
                callback?.onPerformanceUpdate(metrics)
                
                // 重置帧计数
                frameCount.set(0)
                lastFpsCalcTime.set(SystemClock.elapsedRealtime())
                
                scheduleCollection()
            }
        }, sampleIntervalMs)
    }
    
    /**
     * 性能回调接口
     */
    interface PerformanceCallback {
        fun onPerformanceUpdate(metrics: PerformanceMetrics)
    }
}

/**
 * 性能指标数据类
 */
data class PerformanceMetrics(
    val frameRate: Double,
    val droppedFrames: Int,
    val decodedFrames: Int,
    val videoDecodeTimeMs: Double,
    val audioDecodeTimeMs: Double,
    val renderTimeMs: Double,
    val bufferLengthMs: Int,
    val bandwidthBps: Int?,
    val isHardwareDecoding: Boolean,
    val videoBitrate: Int?,
    val audioBitrate: Int?,
    val width: Int?,
    val height: Int?
) {
    fun toMap(): Map<String, Any?> {
        return mapOf(
            "frameRate" to frameRate,
            "droppedFrames" to droppedFrames,
            "decodedFrames" to decodedFrames,
            "videoDecodeTimeMs" to videoDecodeTimeMs,
            "audioDecodeTimeMs" to audioDecodeTimeMs,
            "renderTimeMs" to renderTimeMs,
            "bufferLengthMs" to bufferLengthMs,
            "bandwidthBps" to bandwidthBps,
            "isHardwareDecoding" to isHardwareDecoding,
            "videoBitrate" to videoBitrate,
            "audioBitrate" to audioBitrate,
            "width" to width,
            "height" to height
        )
    }
}
